zkevm-circuits/src/ecc_circuit.rs

```diff
@@ -0,0 +1,688 @@
+                        powers_of_rand.iter().cloned(),
+                    ),
+                    y_c0_rlc: pairing_chip.fp_chip.range.gate.inner_product(
+                        ctx,
+                        y_c0_cells,
+                        powers_of_rand.iter().cloned(),
+                    ),
+                    y_c1_rlc: pairing_chip.fp_chip.range.gate.inner_product(
+                        ctx,
+                        y_c1_cells,
+                        powers_of_rand.iter().cloned(),
+                    ),
+                }
+            })
+            .collect_vec();
+
+        // RLC over the entire input bytes.
+        let input_cells = std::iter::empty()
+            .chain(g1s.iter().map(|g1| g1.decomposed.x_cells.clone()))
+            .chain(g1s.iter().map(|g1| g1.decomposed.y_cells.clone()))
+            .chain(g2s.iter().map(|g2| g2.decomposed.x_c0_cells.clone()))
+            .chain(g2s.iter().map(|g2| g2.decomposed.x_c1_cells.clone()))
+            .chain(g2s.iter().map(|g2| g2.decomposed.y_c0_cells.clone()))
+            .chain(g2s.iter().map(|g2| g2.decomposed.y_c1_cells.clone()))
+            .flatten()
+            .collect::<Vec<QuantumCell<F>>>();
+        let input_rlc = pairing_chip.fp_chip.range.gate.inner_product(
+            ctx,
+            input_cells,
+            powers_of_rand.iter().cloned(),
+        );
+
+        let pairs = g1s
+            .iter()
+            .zip(g2s.iter())
+            .map(|(g1, g2)| (&g1.decomposed.ec_point, &g2.decomposed.ec_point))
+            .collect_vec();
+
+        let success = {
+            let gt = {
+                let gt = pairing_chip.multi_miller_loop(ctx, pairs);
+                fp12_chip.final_exp(ctx, &gt)
+            };
+            // whether pairing check was successful.
+            let one = fp12_chip.load_constant(ctx, Fq12::one());
+            fp12_chip.is_equal(ctx, &gt, &one)
+        };
+
+        ecc_chip.field_chip().range().gate().assert_equal(
+            ctx,
+            QuantumCell::Existing(success),
+            QuantumCell::Witness(Value::known(
+                op.output.to_scalar().expect("EcPairing output = {0, 1}"),
+            )),
+        );
+
+        EcPairingAssigned {
+            g1s,
+            g2s,
+            input_rlc,
+            success,
+        }
+    }
+
+    fn assign_g1(
+        &self,
+        ctx: &mut Context<F>,
+        fp_chip: &EccChip<F, FpConfig<F, Fq>>,
+        g1: G1Affine,
+        powers_of_rand: &[QuantumCell<F>],
+    ) -> G1Assigned<F> {
+        let ec_point = fp_chip.load_private(ctx, (Value::known(g1.x), Value::known(g1.y)));
+        let (x_cells, y_cells) = self.decompose_g1(g1);
+        let decomposed = G1Decomposed {
+            ec_point,
+            x_cells: x_cells.clone(),
+            y_cells: x_cells.clone(),
+        };
+        G1Assigned {
+            decomposed,
+            x_rlc: fp_chip.field_chip().range.gate.inner_product(
```

Comment:
the way that we are doing RLCs may be vulnerable...
For instance, if I have two points `(x1, y1), (x2, y2)` and the RLC challenge `alpha` (ignoring limbs for now)
the right way for RLC would be along the line of  `x1 + alpha y1 + alpha^2 x2 +  alpha^3 y2`

currently it is, iiuc, 
```
(x1+ x2) + (y1 + y2) alpha
```

that means, for a set of k group elements, an attacker can have `k-1` rubbish data, and then use the last one to offset the noise in the previous elements

Reply:
At the moment:
```
x_rlc = x[0].r + x[1].r^2 + ... + x[31].r^31
```
and so on.

And this is later copied via the `copy_advice` method to the ECC table's relevant column `arg_rlc`. The EVM circuit then does a lookup to the ECC table using the RLC value of `P_x`, and so on.